### PR TITLE
feat: Optimize locking for SNMP MIBs loading.

### PR DESCRIPTION
**Pull request by srebhan**
- [ ] Updated associated README.md.
- [ ] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

This PR optimizes locking for the SNMP MIBs loading. This is necessary as the current loading of those files is completely serial. With large setups with dozends or hundreds of SNMP plugins this might take long...

@Hipska will provide measurements... :-)

### CODE REVIEW

**Comment by Hipska** [comment]
### Current release
```log
2021-12-02T14:12:40Z I! Starting Telegraf 1.20.4
2021-12-02T14:12:41Z D! [agent] Initializing plugins
2021-12-02T14:12:41Z D! [agent] Connecting outputs
```
_Note_: This version still used `snmptranslate`  and `snmptable` on the fly when the input is started. (this takes around 10s on a fast system)

### Current master (1.21.0 RC0)
```log
2021-12-02T14:19:43Z I! Starting Telegraf 1.21.0
2021-12-02T14:19:44Z D! [agent] Initializing plugins
```
After this, I waited for 30 minutes until I killed it. CPU was high and memory usage grew steadily to 14GiB during that time.

### Artefact 0e707ebc
```log
2021-12-02T15:05:14Z I! Starting Telegraf 1.21.0-0e707ebc
2021-12-02T15:05:16Z D! [agent] Initializing plugins
```
Similar behaviour as RC0, stopped after 23 minutes and 10GiB of memory usage.

**Comment by srebhan** [comment]
With the added benchmark, i.e. loading the `testdata` directory multiple times I see the following

#### Current master
```text
goos: linux
goarch: amd64
pkg: github.com/influxdata/telegraf/plugins/inputs/snmp
cpu: AMD Ryzen 7 3700X 8-Core Processor             
BenchmarkMibLoading
BenchmarkMibLoading-16                 3         438131129 ns/op
PASS
```

#### Optimized locking
```text
goos: linux
goarch: amd64
pkg: github.com/influxdata/telegraf/plugins/inputs/snmp
cpu: AMD Ryzen 7 3700X 8-Core Processor             
BenchmarkMibLoading
BenchmarkMibLoading-16                 3         452756498 ns/op
PASS
```

#### Caching
```text
goos: linux
goarch: amd64
pkg: github.com/influxdata/telegraf/plugins/inputs/snmp
cpu: AMD Ryzen 7 3700X 8-Core Processor             
BenchmarkMibLoading
BenchmarkMibLoading-16           4237710               288.9 ns/op
PASS
```

**Comment by reimda** [comment]
@hipska Do you see acceptable performance with 336b0ce?

**Comment by Hipska** [comment]
### Artefact 336b0ced
```log
2021-12-03T08:35:15Z I! Starting Telegraf 1.21.0-336b0ced
2021-12-03T08:35:16Z D! [agent] Initializing plugins
2021-12-03T08:35:21Z D! [agent] Connecting outputs
...
```
Whoa, this is what it should have been!

File: internal/snmp/translate.go [hunk 1]
```diff
@@ -40,14 +69,14 @@ func LoadMibsFromPath(paths []string, log telegraf.Logger) error {
...
 			err := filepath.Walk(folder, func(path string, info os.FileInfo, err error) error {
 				// checks if file or directory
 				if info.IsDir() {
-					gosmi.AppendPath(path)
+					appendPath(path)
 				} else if info.Mode()&os.ModeSymlink == 0 {
-					_, err := gosmi.LoadModule(info.Name())
-					if err != nil {
+					if err := loadModule(info.Name()); err != nil {
 						log.Warnf("Module could not be loaded %v", err)
```
Comment (Hipska) [inline internal/snmp/translate.go:80]:
```suggestion
						log.Warn(err)
```
The error from gosmi already has the same/similar text.

**Review by Hipska — APPROVED** [review]
Would you willing to change the log messages a little bit while we are here?

**Review by Hipska — APPROVED** [review]

**Review by MyaLongmire — APPROVED** [review]
Thanks for doing this Sven

**Comment by srebhan** [comment]
@MyaLongmire thanks for making this possible in the first place. That PR was fun. :-)

**Review by reimda — APPROVED** [review]

**Pull request merged**